src/server/pfs/server/driver.go

```diff
@@ -2533,9 +2535,41 @@ func (d *driver) getFile(pachClient *client.APIClient, file *pfs.File, offset in
 		},
 	)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	return grpcutil.NewStreamingBytesReader(getBlocksClient, nil), nil
+	return grpcutil.NewStreamingBytesReader(getBlocksClient, nil), nil, nil
+}
+
+func (d *driver) filesFromByteStream(server pfs.API_GetFilesServer, pachClient *client.APIClient, objReader io.Reader, file *pfs.File, nodes []*hashtree.NodeProto, f func(*pfs.GetFileResponse) error) error {
+	var buf bytes.Buffer
+
+	// If the size of the requested file is larger than max message size,
+	// it must be split into multiple GetFileResponse messages
+	// with only the first containing non-nil File metadata.
+	for _, node := range nodes {
+		for i := int64(0); i <= node.SubtreeSize; i += int64(grpcutil.MaxMsgSize) {
+			_, err := io.CopyN(&buf, objReader, int64(node.SubtreeSize))
+			if err != nil && err != io.EOF {
+				return err
+			}
+			gfr := &pfs.GetFileResponse{
+				Value: buf.Bytes(),
+			}
+			buf.Reset()
+
+			if i == 0 {
+				gfr.File = file
+				// TODO: is this right to use the repo/commit from the request, but the path from the node name?
+				gfr.File.Path = node.GetName()
```

Comment:
One thing I realized is that the file path we receive from the request obviously does not correspond to the file path of the actual file we are getting because the request only has one path (which could be a glob patter or in the future directories + glob patterns). My question here is whether or not we can just assume the name on the NodeProto too be the file path? This works for the naive case, but am unclear on how it will work with directories.  

I added the node hash to the get file response because that seemed like more canonical metadata about a file. I'd imagine we either want one or the other to represent the metadata of a file (File, Hash, or something else), but added both now as a reminder for discussion. Thoughts?

Reply:
There's a function in this file `nodeToFileInfoHeaderFooter` that will give you back a `pfs.FileInfo` with all the relevant information filled in. You could take a look at / just use that. I think for the `Name` field it's just copying it over.

Also I'm not sure if it makes sense to return the `Hash` as part of `GetFiles`, that's part of the metadata for the file which is returned as part of `ListFile` or `InspectFile`. Generally once you're to the point of downloading the files you already know the hash, or if you don't it's because you don't need it for anything.

Reply:
Ya looking at `nodeToFIleInfoHeaderFooter` I do see that it is just copying the `Name` over. Therefore I think it makes the most sense to just leave it as is, because trying to actually call `nodeToFileInfo...()` here gets ugly due to different behavior between input and output repos.